File: app/schemas/learning.py
"""User learning-loop schemas."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class PracticeSessionCreate(BaseModel):
    title: str | None = Field(default=None, max_length=255)
    question_ids: list[UUID] = Field(default_factory=list)
    knowledge_point_id: UUID | None = None
    limit: int = Field(default=10, ge=1, le=50)


class PracticeQuestionView(BaseModel):
    id: UUID
    question_text: str
    question_type: str
    options: list[dict] | None = None
    knowledge_point_ids: list[UUID] = Field(default_factory=list)


class PracticeSessionResponse(BaseModel):
    id: UUID
    title: str
    status: str
    source_type: str | None = None
    source_id: str | None = None
    question_ids: list[UUID] = Field(default_factory=list)
    current_index: int
    total_questions: int
    correct_count: int
    score: float
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    questions: list[PracticeQuestionView] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class PracticeSessionListResponse(BaseModel):
    items: list[PracticeSessionResponse]
    total: int
    page: int
    page_size: int


class PracticeSubmitRequest(BaseModel):
    question_id: UUID
    answer: str = Field(..., min_length=1, max_length=5000)


class PracticeAttemptResponse(BaseModel):
    id: UUID
    session_id: UUID
    question_id: UUID
    submitted_answer: str
    correct_answer: str | None
    is_correct: bool
    explanation: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class PracticeSubmitResponse(BaseModel):
    attempt: PracticeAttemptResponse
    session: PracticeSessionResponse
    wrong_question_updated: bool
    review_card_due_at: datetime | None = None


class WrongQuestionResponse(BaseModel):
    id: UUID
    question_id: UUID
    question_text: str
    wrong_count: int
    last_answer: str | None = None
    resolved_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class WrongQuestionListResponse(BaseModel):
    items: list[WrongQuestionResponse]
    total: int
    page: int
    page_size: int


class MasteryRecordResponse(BaseModel):
    id: UUID
    question_id: UUID | None = None
    knowledge_point_id: UUID | None = None
    mastery_score: float
    attempts_count: int
    correct_count: int
    last_practiced_at: datetime | None = None

    model_config = {"from_attributes": True}


class MasteryRecordListResponse(BaseModel):
    items: list[MasteryRecordResponse]
    total: int


class ReviewCardResponse(BaseModel):
    id: UUID
    question_id: UUID
    question_text: str
    status: str
    due_at: datetime
    interval_days: int
    ease_factor: float
    last_result: str | None = None


class ReviewCardListResponse(BaseModel):
    items: list[ReviewCardResponse]
    total: int
    page: int
    page_size: int


class LearningPathItemResponse(BaseModel):
    id: UUID
    knowledge_point_id: UUID | None = None
    question_id: UUID | None = None
    title: str
    status: str
    priority: int
    reason: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class LearningPathResponse(BaseModel):
    items: list[LearningPathItemResponse]
    total: int

File: front/shared/types/learning.ts
export interface PracticeSessionCreate {
  title?: string | null
  question_ids?: string[]
  knowledge_point_id?: string | null
  limit?: number
}

export interface PracticeQuestionView {
  id: string
  question_text: string
  question_type: string
  options?: Array<{ label: string; text: string }> | null
  knowledge_point_ids: string[]
}

export interface PracticeSessionResponse {
  id: string
  title: string
  status: string
  source_type?: string | null
  source_id?: string | null
  question_ids: string[]
  current_index: number
  total_questions: number
  correct_count: number
  score: number
  completed_at?: string | null
  created_at: string
  updated_at: string
  questions: PracticeQuestionView[]
}

export interface PracticeSessionListResponse {
  items: PracticeSessionResponse[]
  total: number
  page: number
  page_size: number
}

export interface PracticeSubmitRequest {
  question_id: string
  answer: string
}

export interface PracticeAttemptResponse {
  id: string
  session_id: string
  question_id: string
  submitted_answer: string
  correct_answer?: string | null
  is_correct: boolean
  explanation?: string | null
  created_at: string
}

export interface PracticeSubmitResponse {
  attempt: PracticeAttemptResponse
  session: PracticeSessionResponse
  wrong_question_updated: boolean
  review_card_due_at?: string | null
}

export interface WrongQuestionResponse {
  id: string
  question_id: string
  question_text: string
  wrong_count: number
  last_answer?: string | null
  resolved_at?: string | null
  created_at: string
  updated_at: string
}

export interface WrongQuestionListResponse {
  items: WrongQuestionResponse[]
  total: number
  page: number
  page_size: number
}

export interface MasteryRecordResponse {
  id: string
  question_id?: string | null
  knowledge_point_id?: string | null
  mastery_score: number
  attempts_count: number
  correct_count: number
  last_practiced_at?: string | null
}

export interface MasteryRecordListResponse {
  items: MasteryRecordResponse[]
  total: number
}

export interface ReviewCardResponse {
  id: string
  question_id: string
  question_text: string
  status: string
  due_at: string
  interval_days: number
  ease_factor: number
  last_result?: string | null
}

export interface ReviewCardListResponse {
  items: ReviewCardResponse[]
  total: number
  page: number
  page_size: number
}

export interface LearningPathItemResponse {
  id: string
  knowledge_point_id?: string | null
  question_id?: string | null
  title: string
  status: string
  priority: number
  reason?: string | null
  created_at: string
  updated_at: string
}

export interface LearningPathResponse {
  items: LearningPathItemResponse[]
  total: number
}
